### PR TITLE
Streamlined setting environment variables into setup_ckanr, superseding set_test_env

### PR DESCRIPTION
**Pull request by florianm**
Following up on discussion of #31 - first cut - @sckott @wush978 what do you think?

### CODE REVIEW

**Comment by florianm** [comment]
Ok, this is starting to become a major update - thanks to @wush978 things are now clearer and more consistent but in the course of this refactoring I've had to touch `ckanr` in a lot of places.

* `set_ckanr_url()` and `set_api_key()` are gone, superseded by `ckanr_settings()`.
* functions default to `url=get_default_url()` and `key=get_default_key()`.
* `ckan_POST` now expects a key for authenticated requests - no more defaulting to `api_key()` - authenticated functions now use `ckan_POST` and hand over a working key.
* tests consistently use the configured test server.
* some test settings, getters and test helpers added for groups and organizations.
* ISIS mode for tests - blow up or skip - set through `test_behaviour`.
* major formatting spring clean (line length, white spaces in function arguments).

 @sckott e.g. runs tests on my server like so:
```r
# Scott Chamberlain @ data-demo
setup_ckanr(test_url="http://data-demo.dpaw.wa.gov.au/",
            test_key="scotts-test-key",
            test_did="ae6fdf6f-1435-444b-b319-198cc84e4594",
            test_rid="0155c395-842d-4b66-a72c-c6e688b260c3",
            test_gid="templates-and-guidelines",
            test_oid="sandbox",
            test_behaviour="FAIL")
devtools::test()
```

Looking forward to constructive feedback!

File: R/ds_search_sql.R [hunk 1]
```diff
@@ -12,8 +12,10 @@
...
 
-ds_search_sql <- function(sql, url = 'http://demo.ckan.org', as='list', ...) {
-  res <- POST(file.path(url, '/api/action/datastore_search_sql'), ctj(), query = list(sql = sql), ...)
+ds_search_sql <- function(sql, url=get_default_url(), as='list', ...) {
+  res <- POST(file.path(url, '/api/action/datastore_search_sql'),
+              ctj(),
+              query=list(sql = sql), ...)
   res <- content(res, "text")
-  switch(as, json = res, list = jsl(res), table = jsd(res))
+  switch(as, json=res, list=jsl(res), table=jsd(res))
```
Comment (sckott) [inline R/ds_search_sql.R]: @florianm can you undo these changes to spacing around `=` signs? I think many agree that code is more readable with spaces in appropriate places

File: R/package_activity_list.R [hunk 2]
```diff
@@ -9,10 +9,12 @@
 #' package_activity_list("34d60b13-1fd5-430e-b0ec-c8bc7f4841cf", as="table")
 #' package_activity_list("34d60b13-1fd5-430e-b0ec-c8bc7f4841cf", as="json")
 #' }
-package_activity_list <- function(id, offset = 0, limit = 31, url = get_ckanr_url(),
+package_activity_list <- function(id,
+                                  offset = 0,
+                                  limit = 31,
+                                  url=get_default_url(),
                                   as="list", ...) {
```
Comment (sckott) [inline R/package_activity_list.R]: @florianm Can you undo these newlines between parameters.

File: tests/testthat/helper-ckanr.R [hunk 3]
```diff
@@ -1,25 +1,72 @@
...
+#'
+#' @keywords internal
+#' @importFrom testthat skip
+#' @details
+#' \code{check_ckan} will allow a test to fail (and not hide error messages)
+#' if \code{setup_ckanr(test_strict="TRUE")} was set.
+#' \code{check_ckan} will allow a test to skip (and the test suite to pass)
+#' if \code{setup_ckanr(test_strict)} was set to anything but the string "TRUE".
+#' @param url A URL that shall be tested whether it is online
+#'
```
Comment (sckott) [inline tests/testthat/helper-ckanr.R]: @florianm You shouldn't need any docstring for these test helper functions. You can do comments `#` if you want. AFAIK we shouldn't need to say we want to import `testthat::skip`

or do you mean for these functions to be available for use in the package via `:::`?

**Comment by sckott** [comment]
In the future, smaller pull requests that are focused around a single theme, would be easier to manage. 

I made a comment about white spaces and parameter newlines. I'd like it if you could undo those changes. We generally follow http://adv-r.had.co.nz/Style.html 

Thoughts @wush978

**Comment by wush978** [comment]
I check this PR on windows but it fails:

```
==> devtools::test()

Loading ckanr
Testing ckanr
changes : 12...
ckanr_settings : .................
ds_create_dataset : ...34
ds_search : 56
ds_search_sql : 78
group_list : 9abc
changes : defg
license_list : h
organization_list : Error in function (type, msg, asError = TRUE)  : <url> malformed
Calls: <Anonymous> ... perform -> <Anonymous> -> .Call -> <Anonymous> -> fun
```

A minor concern:

I do not suggest to use:

```r
 url <- get_test_url()
 key <- get_test_key()
 rid <- get_test_rid()
resource_update(rid, path=path, url=url, key=key)
```

Because this kind of test does not coverage the real use case:

```r
setup_ckanr(url="http://data.techno-science.ca/", key="my-ckan-api-key")
resource_update(rid, path=path)
```

File: R/ckanr_settings.R [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+#' setup_ckanr(url="http://data.techno-science.ca/", key="some-CKAN-API-key",
+#'            test_url="http://test-ckan.gov/",test_key="test-ckan-API-key",
+#'            test_did="test-ckan-dataset-id",test_rid="test-ckan-resource-id",
+#'            test_gid="test-group-name", test_oid="test-organzation-name",
+#'            test_behaviour="FAIL")
+#'
+#' # Not specifying the default CKAN URL will reset the CKAN URL to its default
+#' # "http://data.techno-science.ca/":
+#' setup_ckanr()
+setup_ckanr <- function(
```
Comment (sckott) [inline R/ckanr_settings.R]: @florianm I would suggest making this function to set settings the same name format as the function to get settings: So `ckanr_setup` and `ckanr_settings`, that okay?

**Comment by sckott** [comment]
@wush978 you have to set a few env vars - see https://github.com/ropensci/ckanr/pull/37#issuecomment-104535177

After setting those, tests did pass for me locally

**Comment by sckott** [comment]
@florianm regarding @wush978 point  about `get_*()` functions, is the purpose of those functions to be able to use in the test suite so that we, or other users, can pass in the url/key/etc. as a parameter instead of functions reading env vars after using `setup_ckanr()`?

**Comment by florianm** [comment]
Hi all,

Thanks for the feedback, I'll get to incorporate all of your requests
hopefully on Monday! They all make sense.
With the latest status quo, do we need a separate test url and key, or
should tests set and implicitly/explicitly use the default url and key?

Cheers and thanks for the review,
Florian

File: R/ckanr_settings.R (hunk 4 above)
Comment (florianm) [inline R/ckanr_settings.R]: That would be even more consistent, will do!

Comment (sckott) [inline R/ckanr_settings.R]: great :+1:

**Comment by sckott** [comment]
@florianm  wrt putting each param on a separate line - I imagine you did so so it's easier to run default params to work on the function?  I totally understand that. That's why I use this https://github.com/sckott/sacbox/blob/master/R/load_defaults.r  I just load that on every R session, so that I can load default params for a function I'm working on, and don't need to run one at a time.

**Comment by sckott** [comment]
> With the latest status quo, do we need a separate test url and key, or should tests set and implicitly/explicitly use the default url and key?

I guess CKAN is a little different from other things in that a user could potentially be a developer/data provider and want to test against different servers with different auth settings. So maybe it makes sense to set different settings for test server

File: R/ds_search_sql.R (hunk 1 above)
Comment (florianm) [inline R/ds_search_sql.R]: Yelp... OK I'l bring them back! I went with the Google R style guide which finds [spaces being optional for function parameters](https://google-styleguide.googlecode.com/svn/trunk/Rguide.xml#spacing). Agree with Hadley's "always spaces around infix operators".

File: R/package_activity_list.R (hunk 2 above)
Comment (florianm) [inline R/package_activity_list.R]: OK, will undo. I did this to parameter lists longer than one line. I'll reflow them but still line break them at 80 chars if that's OK.

File: tests/testthat/helper-ckanr.R (hunk 3 above)
Comment (florianm) [inline tests/testthat/helper-ckanr.R]: I probably should have used inline comments then. Just wanted to document this function to you and Wush - what's the best formal way to document internal functions?

**Comment by florianm** [comment]
Thanks for the feedback, good points!

* re PR size - apart from formatting (which I'll undo), as this PR changes some core workflows (and @wush978 rightly wouldn't let me get away with half-way solutions) I found no other way than going all in.

* re param line breaks - to my Python-based taste, readability outweighs brevity - lists of parameters passed to functions longer than 80 chars looks better when line-fed individually. It's easier for caffeine-deprived developers like me to spot omissions while coding (especially when adding/removing parameters), and to end-users those line breaks will be hidden. But that's only my personal taste and as you suggested, I'll condense the parameters again.

* re parameter=default spacing - see above, went with Google style, will convert to Church of Hadley. Law, order and aesthetically pleasing spacing is only one grep/xargs/sed away.

* re default vs test url and key - with my initial cut, having separate default and test settings made sense to me. But @wush978 has a valid point in wanting tests to use the default url and key implicitly. If we can assume that testers will run `ckanr_setup` before running the tests anyways, and not use the same ckanr instance against a production server while the tests are running (right?), then we won't need separate url and key, and we can simplify the parameter names. So end-users would still run

```r
ckanr_setup(url="sdfsdf", key="sdfsf")
```

while testers would run (if we simplify the parameter names):
```r
ckanr_setup(url="test ckan", key="test key", did="test ds id", rid="test res id", gid="test group name",
oid="test org id", test_behaviour="FAIL")

# testing whether function defaults to ckanr_settings correctly:
x <- do_something() # defaulting to get_{url, key, did, rid, gid, oid}()

# testing with explicit parameters
y <- do_something(url=get_url(), key=get_key(), did=get_did(), rid=get_rid(), gid=get_gid(), oid=get_oid())
```

* re `load_defaults.R` - nice. I will totally steal this - but setting defaults like this was neither within my skill set nor my reasoning :-)

So, last question to you @sckott and @wush978 :

Should I keep DEFAULT and TEST url and key separate, or merge and simplify?

**Comment by florianm** [comment]
Formatting clean-up is done, tests pass with Scott's settings on data-demo.dpaw.wa.gov.au. Pending my actions on whether to keep or merge default and test url/key, this PR is nearing completion.

**Comment by sckott** [comment]
Thanks @florianm 

> Should I keep DEFAULT and TEST url and key separate, or merge and simplify? If we merge, what parameter names would be the best (should I drop the "test_")? Is there a shorter name for "test_behaviour"?

@wush978 do you have an opinion on this?

**Comment by wush978** [comment]
Sorry that I cannot closely check these codes. I will check them on the weekend.

**Comment by sckott** [comment]
@florianm I think this is good. Mind if I merge now?

**Comment by florianm** [comment]
All good from my side, and if you and Wush feel merging default and test url/key would be appropriate I can still do so in a new PR.

**Comment by sckott** [comment]
Yeah, I'd like to get this merged - and we can make additional changes after merging

**Comment by wush978** [comment]
Dear @florianm , 

I think dropping `test_` is ok. However, it is not an urgent issue, so maybe we could change it in the future.

**Comment by florianm** [comment]
Thanks @wush978 and @sckott - let me know if you wish to merge and I'll submit another PR.

**Comment by sckott** [comment]
@florianm merge what? this is already merged

**Comment by florianm** [comment]
Sorry - I meant url and key default and test settings (along [these lines](https://github.com/ropensci/ckanr/pull/37#issuecomment-105100427))!

**Comment by sckott** [comment]
@florianm yes, plz do. 

regarding style - in ropensci pkgs we try to maintain a similar style across pkgs for consistency within ropensci. totally fine with other styles, but we just want to maintain consistency within our suite as much as possible
